src/transformers/models/llama/modeling_llama.py

```diff
@@ -388,11 +413,189 @@ def forward(
         return attn_output, attn_weights, past_key_value
 
 
+class LlamaFlashAttention2(LlamaAttention):
+    """
+    Llama flash attention module. This module inherits from `LlamaAttention` as the weights of the module stays
+    untouched. The only required change would be on the forward pass where it needs to correctly call the public API of
+    flash attention and deal with padding tokens in case the input contains any of them.
+    """
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value: Optional[Tuple[torch.Tensor]] = None,
+        output_attentions: bool = False,
+        use_cache: bool = False,
+        padding_mask: Optional[torch.LongTensor] = None,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
+        # LlamaFlashAttention2 attention does not support output_attentions
+        output_attentions = False
+
+        bsz, q_len, _ = hidden_states.size()
+
+        query_states = self.q_proj(hidden_states)
+        key_states = self.k_proj(hidden_states)
+        value_states = self.v_proj(hidden_states)
+
+        # Flash attention requires the input to have the shape
+        # batch_size x seq_length x head_dime x hidden_dim
+        # therefore we just need to keep the original shape
+        query_states = query_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
+        key_states = key_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
+        value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
+
+        kv_seq_len = key_states.shape[-2]
+        if past_key_value is not None:
+            kv_seq_len += past_key_value[0].shape[-2]
+
+        cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
+
+        query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
+
+        if past_key_value is not None:
+            # reuse k, v, self_attention
+            key_states = torch.cat([past_key_value[0], key_states], dim=2)
+            value_states = torch.cat([past_key_value[1], value_states], dim=2)
+
+        past_key_value = (key_states, value_states) if use_cache else None
+
+        query_states = query_states.transpose(1, 2)
+        key_states = key_states.transpose(1, 2)
+        value_states = value_states.transpose(1, 2)
+
+        # TODO: llama does not have dropout in the config??
+        # It is recommended to use dropout with FA according to the docs
+        # when training.
+        dropout_rate = 0.0  # if not self.training else self.attn_dropout
+
+        # In PEFT, usually we cast the layer norms in float32 for training stability reasons
+        # therefore the input hidden states gets silently casted in float32. Hence, we need
```

Comment:
The hidden states are cast in float32 for a fp32 layer norm weight here:

https://github.com/huggingface/transformers/blob/5936c8c57ccb2bda3b3f28856a7ef992c5c9f451/src/transformers/models/llama/modeling_llama.py#L113

How about casting the output of the layer norm in the same type as the input tensor?

```python
return (self.weight * hidden_states).to(input_dtype)
````

Reply:
I think this would work too, let me try to run few experiments and get back to you

Reply:
This PR: https://github.com/huggingface/transformers/pull/24891/files#diff-06392bad3b9e97be9ade60d4ac46f73b6809388f4d507c2ba1384ab872711c51 re-introduced the bug that was fixed in https://github.com/huggingface/transformers/pull/23535

Reply:
I think this was intended but I am not sure, I prefer to wait for @ArthurZucker to have a second opinion here

Reply:
This is just the 1 to 1 function used in Llama origiinal code

Reply:
And it's mathematically not the same computation if I am not mistaken.

Reply:
Consider we have fp16 layer norm weight:
Above: (fp16 * fp16) -> fp16    Below: (fp16 * fp32) -> fp16
If we have fp32 layer norm weight:
Above: (fp32 * fp16) -> fp32    Below: (fp32 * fp32) -> fp16

Is this right?

Reply:
It is, but if you run the following:
```python 
import torch

hidden_states = torch.rand(4096, 4096, dtype = torch.float32)
weights = torch.rand(4096, dtype = torch.float32)

for i in [1e-2,1e-3,1e-4,1e-5]:
    try:
        torch.testing.assert_close(weights * hidden_states, (weights * hidden_states.half()), rtol=i, atol=i)

    except:
        print(i, "found diff in float32")

    try:
        torch.testing.assert_close((weights * hidden_states).half(), (weights * hidden_states.half()).half(), rtol=i, atol=i)
    except :
        print(i, "found diff in float16")
``` 
you'll see that neither in float16 nor in float32 does this pass for `1e-4` and `1e-5` which means potential propagation. When porting the model this was an issue. It's mathematically not the same computation.

Reply:
I get it, you mean that if we cast both weights and hidden states in fp16 precision, there will be a large divergence in the results which is unacceptable, so we can only cast either weights or hidden states in fp16 precision.

Reply:
@younesbelkada  Sorry for disturbing you again, should we use the Bfloat16 data type in bf16 training?
https://github.com/huggingface/transformers/blob/3ca18d6d09ee0d1610a400ead6f6041394f66421/src/transformers/models/llama/modeling_llama.py#L485-L487

Reply:
no worries; 
Yes I think that it makes sense, it might be challenging to retrieve the correct dtype, let me dig into that and look into it. Would you mind opening a new ticket for it and tag me there? 🙏 Thanks!